odl/space/fspace.py

```diff
@@ -25,51 +25,254 @@
+    True
+    >>> mg_dense = np.meshgrid(x, y, indexing='ij', sparse=False)
+    >>> np.array_equal(step(mg_dense), step_vec(mg_dense))
+    True
+
+    With output parameter:
+
+    >>> @vectorize(dtype=float, outarg='positional')
+    ... def step(x):
+    ...     return 0 if x[0] <= 0 else 1
+    >>> x = np.linspace(-5, 13, 10, dtype=float).reshape((2, 5))
+    >>> out = np.empty(5, dtype=float)
+    >>> step(x, out)  # returns out
+    array([ 0.,  0.,  0.,  1.,  1.])
+    """
+    def vect_decorator(func):
+        def _vect_wrapper(x, out, **kwargs):
+            if isinstance(x, np.ndarray):  # array
+                if x.ndim == 1:
+                    dim = 1
+                elif x.ndim == 2:
+                    dim = len(x)
+                    if dim == 1:
+                        x = x.squeeze()
+                else:
+                    raise ValueError('only 1- or 2-dimensional arrays '
+                                     'supported.')
+            else:  # meshgrid
+                dim = len(x)
+                if dim == 1:
+                    x = x[0]
+            if dim == 1:
+                if out is None:
+                    out = np.empty(x.size, dtype=dtype)
+                for i, xi in enumerate(x):
+                    out[i] = func(xi, **kwargs)
+                return out
+            else:
+                if is_valid_input_array(x, dim):
+                    if out is None:
+                        out = np.empty(x.shape[1], dtype=dtype)
+                    for i, pt in enumerate(x.T):
+                        out[i] = func(pt, **kwargs)
+                    return out
+                elif is_valid_input_meshgrid(x, dim):
+                    order = meshgrid_input_order(x)
+                    if out is None:
+                        out_shape = np.broadcast(*x).shape
+                        out = np.empty(out_shape, dtype=dtype, order=order)
+
+                    # TODO: find a better way to iterate
+                    vecs = vecs_from_meshgrid(x, order=order)
+                    for i, pt in enumerate(product(*vecs)):
+                        out.flat[i] = func(pt, **kwargs)
+                    return out
+                else:
+                    if out is None:
+                        try:
+                            return func(x)
+                        except (TypeError, ValueError, IndexError):
+                            raise TypeError('invalid vectorized input type.')
+                    else:
+                        raise TypeError('invalid vectorized input type.')
+
+        @wraps(func)
+        def vect_wrapper_no_out(x, **kwargs):
+            if 'out' in kwargs:
+                raise TypeError("{}() got an unexpected keyword 'out'."
+                                "".format(func.__name__))
+            return _vect_wrapper(x, None, **kwargs)
+
+        @wraps(func)
+        def vect_wrapper_pos_out(x, out, **kwargs):
+            if out is None:
```

Comment:
Need to look into this. The thing is that if one uses positional `out` and provides `out=None`, the `_vect_wrapper` thinks that we evaluate out-of-place, and it fails in weird ways or, worse, does not fail at all. The wrapping `__call__()` should still fail, but I remember getting confused by this and wanting to print a meaningful message. Maybe there's a better way to do this.